modules/dbparser/patterndb.c

```diff
@@ -673,13 +672,23 @@ _pattern_db_process_matching_rule(PatternDB *self, PDBProcessParams *process_par
   g_string_free(buffer, TRUE);
 }
 
+static void
+_pattern_db_advance_time_and_flush_expired(PatternDB *self, LogMessage *msg)
+{
+  PDBProcessParams process_params = {0};
+
+  g_static_rw_lock_writer_lock(&self->lock);
+  _advance_time_based_on_message(self, &process_params, &msg->timestamps[LM_TS_STAMP]);
+  g_static_rw_lock_writer_unlock(&self->lock);
+  _flush_emitted_messages(self, &process_params);
+}
+
 static void
 _pattern_db_process_unmatching_rule(PatternDB *self, PDBProcessParams *process_params)
 {
   LogMessage *msg = process_params->msg;
 
   g_static_rw_lock_writer_lock(&self->lock);
```

Comment:
After the extraction of `_pattern_db_advance_time_and_flush_expired()`, this lock became unnecessary as `_emit_message()` operates only on `process_params` (per-thread, stack variable), and uses only a read-only variable from `self`.